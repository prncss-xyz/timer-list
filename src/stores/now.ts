import { atom } from "jotai";

const resolution = 200;

export const nowAtom = atom(0);
nowAtom.onMount = (setAtom) => {
  const timer = setInterval(() => {
    return setAtom(Date.now());
  }, resolution);
  return () => clearInterval(timer);
};
