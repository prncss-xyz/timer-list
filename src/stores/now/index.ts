import { atom } from "jotai";

export const nowAtom = atom(0);
