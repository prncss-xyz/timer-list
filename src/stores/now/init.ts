import { useSetAtom } from "jotai";
import { useEffect } from "react";

import { nowAtom } from ".";

const resolution = 200;

export function useInitNow() {
  const setAtom = useSetAtom(nowAtom);
  useEffect(() => {
    const timer = setInterval(() => {
      return setAtom(Date.now());
    }, resolution);
    return () => clearInterval(timer);
  }, [setAtom]);
  return true;
}
