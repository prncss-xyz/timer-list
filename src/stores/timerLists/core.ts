import { TimerList } from "./model";

import { insert } from "@/utils/arrays";
import { getUUID } from "@/utils/uuid";

export function duplicateId(id: string) {
  return function (timerList: TimerList): TimerList {
    let { items } = timerList;
    const index = items.findIndex((item) => item.id === id);
    if (index < 0) return timerList;
    const item = { ...items[index], id: getUUID() };
    items = insert(items, index + 1, item);
    timerList = { ...timerList, items };
    if (timerList.active === id) timerList = nextActiveItem(timerList);
    return timerList;
  };
}

export function removeId(id: string) {
  return function (timerList: TimerList): TimerList {
    let { active, items } = timerList;
    if (active === id) {
      let index = items.findIndex((item) => item.id === id);
      if (index + 1 === items.length) index = Math.max(0, index - 1);
      else index = index + 1;
      active = items.at(index)?.id ?? "";
    }
    return {
      ...timerList,
      active,
      items: timerList.items.filter((item) => item.id !== id),
    };
  };
}

export function nextActiveItem(timerList: TimerList): TimerList {
  const { active, items } = timerList;
  let index = items.findIndex((item) => item.id === active);
  index = Math.min(index + 1, items.length - 1);
  return { ...timerList, active: items[index].id };
}
