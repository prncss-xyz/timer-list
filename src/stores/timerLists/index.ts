// we use the plural form in filename as it will eventually manage many lisits
import { atom } from "jotai";
import { atomEffect } from "jotai-effect";
import { focusAtom } from "jotai-optics";

import { Item, normalize } from "./model";

import { definedAtom } from "@/utils/definedAtom";
import { fromSeconds, toSeconds } from "@/utils/seconds";
import { getUUID } from "@/utils/uuid";

/* const rawTimerListAtom = atom(normalize({ index: 0, items: [] })); */
const rawTimerListAtom = atom(
  normalize({
    index: 0,
    items: [
      { seconds: 1, id: getUUID() },
      { seconds: 2, id: getUUID() },
      { seconds: 3, id: getUUID() },
    ],
  }),
);
export const timerListAtom = focusAtom(rawTimerListAtom, (o) =>
  o.rewrite(normalize),
);

export const currentIndexAtom = focusAtom(timerListAtom, (o) =>
  o.prop("index"),
);

export const itemsAtom = focusAtom(timerListAtom, (o) => o.prop("items"));

const currentItemAtom = atom(
  (get) => {
    const { index, items } = get(timerListAtom);
    return items[index];
  },
  (get, set, item: Item) => {
    const lists = get(timerListAtom);
    let { index, items } = lists;
    items = items.toSpliced(index, 1, item);
    set(timerListAtom, { ...lists, items });
  },
);

export const currentIdAtom = atom(
  (get) => {
    return get(currentItemAtom).id;
  },
  (get, set, id: string) => {
    const lists = get(timerListAtom);
    const { items } = lists;
    const index = items.findIndex((item) => item.id === id);
    set(timerListAtom, { ...lists, index, items });
  },
);

export const currentSecondsAtom = focusAtom(currentItemAtom, (o) =>
  o.prop("seconds"),
);

export const getIdItemAtom = (id: string) =>
  definedAtom(
    focusAtom(timerListAtom, (o) =>
      o.prop("items").find((item) => item.id === id),
    ),
    { id: "", seconds: 0 },
  );

export const getIdItemSecondsAtom = (id: string) =>
  focusAtom(getIdItemAtom(id), (o) => o.prop("seconds"));

export const getIdItemSecondsTextAtom = (id: string) =>
  focusAtom(getIdItemSecondsAtom(id), (o) => o.iso(fromSeconds, toSeconds));

export const duplicateIdAtom = atom(null, (get, set, id: string) => {
  const lists = get(timerListAtom);
  let { items, index: currentIndex } = lists;
  const index = items.findIndex((item) => item.id === id);
  if (index < 0) return;
  const item = { ...items[index], id: getUUID() };
  items = items.toSpliced(index, 0, item);
  if (index <= currentIndex) currentIndex++;
  set(timerListAtom, { ...lists, index: currentIndex, items });
});

export const removeIdAtom = atom(null, (get, set, id: string) => {
  const lists = get(timerListAtom);
  let { items, index: currentIndex } = lists;
  const index = items.findIndex((item) => item.id === id);
  if (index < 0) return;
  items = items.toSpliced(index, 1);
  if (index < currentIndex) currentIndex--;
  set(timerListAtom, { ...lists, index: currentIndex, items });
});

export const clearItemsAtom = atom(null, (_get, set) => {
  set(itemsAtom, []);
});

export const getTimerUpdateEffect = (cb: () => void) =>
  atomEffect((get) => {
    get(currentSecondsAtom);
    get(currentIdAtom);
    cb();
  });
