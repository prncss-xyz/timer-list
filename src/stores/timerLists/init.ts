import { useAtomValue } from "jotai";
import { useMemo } from "react";

import { getTimerUpdateEffect } from ".";

export function useInitTimerList(cb: () => void) {
  const effect = useMemo(() => getTimerUpdateEffect(cb), [cb]);
  useAtomValue(effect);
}
