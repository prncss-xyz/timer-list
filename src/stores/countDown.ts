import { atom, useAtomValue } from "jotai";
import { atomEffect } from "jotai-effect";
import { useMemo } from "react";

import { currentIndexAtom, currentSecondsAtom } from "./timerLists";
import {
  resetTimerAtom,
  stopTimerAtom,
  timerActiveAtom,
  timerElapsedAtom,
} from "./timers";

import { fromSeconds } from "@/utils/seconds";

const countDownMsAtom = atom((get) => {
  const elapsed = get(timerElapsedAtom);
  const total = get(currentSecondsAtom) * 1000;
  return Math.max(0, total - elapsed);
});

export const countDownSecondsAtom = atom((get) =>
  Math.ceil(get(countDownMsAtom) / 1000),
);

export const getAlarmEffect = (alarm: () => void) =>
  atomEffect((get, set) => {
    if (get(timerActiveAtom) === false || get(countDownMsAtom) > 0) return;
    set(stopTimerAtom);
    set(currentIndexAtom, (index) => index + 1);
    set(resetTimerAtom);
    alarm();
  });

export function useInitCountDown(alarm: () => void) {
  const effect = useMemo(() => getAlarmEffect(alarm), [alarm]);
  useAtomValue(effect);
}

export const countDownTextAtom = atom((get) =>
  fromSeconds(get(countDownSecondsAtom)),
);
