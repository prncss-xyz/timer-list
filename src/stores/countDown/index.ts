import { atom } from "jotai";
import { atomEffect } from "jotai-effect";

import { activeSecondsAtom, nextActiveItemAtom } from "../timerLists";
import { timerElapsedAtom, stopTimerAtom, resetTimerAtom } from "../timers";

import { fromSeconds } from "@/utils/seconds";

function getCountDown(elapsedMs: number, total: number | undefined) {
  if (total === undefined) return 0;
  return Math.max(0, Math.ceil(total - elapsedMs / 1000));
}

const countDownAtom = atom((get) => {
  return getCountDown(get(timerElapsedAtom), get(activeSecondsAtom));
});

export const countDownTextAtom = atom((get) => fromSeconds(get(countDownAtom)));

export const getAlarmEffect = (alarm: () => void) =>
  atomEffect((get, set) => {
    if (get(countDownAtom) > 0) return;
    set(stopTimerAtom);
    set(nextActiveItemAtom);
    set(resetTimerAtom);
    alarm();
  });
