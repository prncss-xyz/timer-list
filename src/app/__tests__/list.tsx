import {
  act,
  fireEvent,
  render,
  screen,
  within,
} from "@testing-library/react-native";
import { createStore, Provider } from "jotai";

import Page from "@/app/list";
import { nowAtom } from "@/stores/now";
import { timerListAtom } from "@/stores/timerLists";

describe("list page", () => {
  it("should render properly", () => {
    const store = createStore();
    store.set(timerListAtom, {
      index: 0,
      items: [
        { seconds: 2, id: "a" },
        { seconds: 3, id: "b" },
      ],
    });
    const tree = render(
      <Provider store={store}>
        <Page />
      </Provider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it("should decrease count when playing", () => {
    const store = createStore();
    store.set(timerListAtom, {
      index: 0,
      items: [
        { seconds: 2, id: "a" },
        { seconds: 3, id: "b" },
      ],
    });
    render(
      <Provider store={store}>
        <Page />
      </Provider>,
    );
    // start playing
    fireEvent.press(screen.getByLabelText("play"));
    act(() => {
      store.set(nowAtom, (now) => now + 1000);
    });
    within(screen.getByLabelText("countdown")).getByText("00:00:01");
    // stop playing
    fireEvent.press(screen.getByLabelText("pause"));
    act(() => {
      store.set(nowAtom, (now) => now + 1000);
    });
    within(screen.getByLabelText("countdown")).getByText("00:00:01");
    fireEvent.press(screen.getByLabelText("play"));
    act(() => {
      store.set(nowAtom, (now) => now + 1000);
    });
  });
  it("should clear list", () => {
    const store = createStore();
    store.set(timerListAtom, {
      index: 0,
      items: [
        { seconds: 1, id: "a" },
        { seconds: 2, id: "b" },
        { seconds: 3, id: "c" },
      ],
    });
    render(
      <Provider store={store}>
        <Page />
      </Provider>,
    );
    act(() => {
      fireEvent.press(screen.getByText("clear all"));
    });
    const res = screen.getAllByLabelText("duration");
    expect(res.length).toBe(1);
    expect(within(res[0]).getByText("00:00:00"));
  });
  it("should duplicate an item", () => {
    const store = createStore();
    store.set(timerListAtom, {
      index: 0,
      items: [
        { seconds: 1, id: "a" },
        { seconds: 2, id: "b" },
        { seconds: 3, id: "c" },
      ],
    });
    render(
      <Provider store={store}>
        <Page />
      </Provider>,
    );
    act(() => {
      fireEvent.press(screen.getAllByLabelText("duplicate")[0]);
    });
    const res = screen.getAllByLabelText("duration");
    expect(res.length).toBe(4);
    expect(within(res[0]).getByText("00:00:01"));
    expect(within(res[1]).getByText("00:00:01"));
    expect(within(res[2]).getByText("00:00:02"));
    expect(within(res[3]).getByText("00:00:03"));
  });
  it("remove add an item", () => {
    const store = createStore();
    store.set(timerListAtom, {
      index: 0,
      items: [
        { seconds: 1, id: "a" },
        { seconds: 2, id: "b" },
        { seconds: 3, id: "c" },
      ],
    });
    render(
      <Provider store={store}>
        <Page />
      </Provider>,
    );
    act(() => {
      fireEvent.press(screen.getAllByLabelText("remove")[0]);
    });
    const res = screen.getAllByLabelText("duration");
    expect(res.length).toBe(2);
    expect(within(res[0]).getByText("00:00:02"));
    expect(within(res[1]).getByText("00:00:03"));
  });
});
