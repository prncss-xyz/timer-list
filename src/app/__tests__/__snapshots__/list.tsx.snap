// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list page should render properly 1`] = `
[
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityLabel="play"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "height": 35,
          "width": 35,
        }
      }
    >
      <Text />
    </View>
    <View
      accessibilityLabel="countdown"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        style={
          [
            {
              "borderColor": "#3DD68C",
              "color": "#3DD68C",
            },
            {
              "borderStyle": "solid",
              "borderWidth": 1,
              "justifyContent": "center",
              "padding": 10,
              "textAlign": "center",
            },
            {
              "fontFamily": "RobotoMono-Medium",
              "fontWeight": "500",
            },
            {
              "fontSize": 28,
            },
          ]
        }
      >
        00:00:02
      </Text>
    </View>
    <View
      accessibilityLabel="reset"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "height": 35,
          "width": 35,
        }
      }
    >
      <Text />
    </View>
  </View>,
  <View
    style={
      {
        "borderBottomWidth": 1,
        "borderColor": "#0090FF",
        "borderStyle": "solid",
        "height": 35,
      }
    }
  />,
  <RCTScrollView
    CellRendererComponent={[Function]}
    ItemSeparatorComponent={[Function]}
    contentContainerStyle={{}}
    data={
      [
        {
          "id": "a",
          "seconds": 2,
        },
        {
          "id": "b",
          "seconds": 3,
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        animatedStyle={
          {
            "value": {},
          }
        }
        cellKey="a"
        collapsable={false}
        entering={[Function]}
        exiting={[Function]}
        index={0}
        item={
          {
            "id": "a",
            "seconds": 2,
          }
        }
        layout={[Function]}
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          {
            "width": "100%",
          }
        }
      >
        <View
          aria-label="current"
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingBottom": 10,
              "paddingTop": 10,
            }
          }
        >
          <View
            accessibilityLabel="duration"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <Text
              style={
                [
                  {
                    "color": "#3DD68C",
                  },
                  {
                    "alignItems": "flex-start",
                    "justifyContent": "center",
                    "padding": 5,
                  },
                  {
                    "fontFamily": "RobotoMono-Medium",
                    "fontWeight": "500",
                  },
                  {
                    "fontSize": 28,
                  },
                ]
              }
            >
              00:00:02
            </Text>
          </View>
          <View
            style={
              {
                "flexDirection": "row",
                "gap": 15,
              }
            }
          >
            <View
              accessibilityLabel="duplicate"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "height": 35,
                  "width": 35,
                }
              }
            >
              <Text />
            </View>
            <View
              accessibilityLabel="remove"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "height": 35,
                  "width": 35,
                }
              }
            >
              <Text />
            </View>
            <View
              accessibilityLabel="edit"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "height": 35,
                  "width": 35,
                }
              }
            >
              <Text />
            </View>
          </View>
        </View>
        <View
          style={
            {
              "borderBottomWidth": 1,
              "borderColor": "#0090FF",
              "borderWidth": 0,
              "width": "100%",
            }
          }
        />
      </View>
      <View
        animatedStyle={
          {
            "value": {},
          }
        }
        cellKey="b"
        collapsable={false}
        entering={[Function]}
        exiting={[Function]}
        index={1}
        item={
          {
            "id": "b",
            "seconds": 3,
          }
        }
        layout={[Function]}
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          {
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingBottom": 10,
              "paddingTop": 10,
            }
          }
        >
          <View
            accessibilityLabel="duration"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <Text
              style={
                [
                  {
                    "color": "#0090FF",
                  },
                  {
                    "alignItems": "flex-start",
                    "justifyContent": "center",
                    "padding": 5,
                  },
                  {
                    "fontFamily": "RobotoMono-Medium",
                    "fontWeight": "500",
                  },
                  {
                    "fontSize": 28,
                  },
                ]
              }
            >
              00:00:03
            </Text>
          </View>
          <View
            style={
              {
                "flexDirection": "row",
                "gap": 15,
              }
            }
          >
            <View
              accessibilityLabel="duplicate"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "height": 35,
                  "width": 35,
                }
              }
            >
              <Text />
            </View>
            <View
              accessibilityLabel="remove"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "height": 35,
                  "width": 35,
                }
              }
            >
              <Text />
            </View>
            <View
              accessibilityLabel="edit"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "height": 35,
                  "width": 35,
                }
              }
            >
              <Text />
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>,
  <View
    accessibilityLabel="clear all"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      style={
        [
          {
            "borderWidth": 1,
            "fontFamily": "Inter-SemiBold",
            "fontSize": 20,
            "fontWeight": "600",
            "padding": 5,
            "textAlign": "center",
            "textTransform": "uppercase",
          },
          {
            "borderColor": "#FFC53D",
            "color": "#FFC53D",
          },
        ]
      }
    >
      clear all
    </Text>
  </View>,
]
`;
